src/services/mapCode.ts

```diff
@@ -0,0 +1,365 @@
+    ];
+    const parsedNodes = [];
+    for (const { parse, body } of nodeKinds) {
+        const sourceFile = parse();
+        if (sourceFile.parseDiagnostics.length === 0) {
+            // If we run into a case with no parse errors, this is likely the right kind.
+            return body(sourceFile);
+        }
+        else {
+            // Otherwise, we'll need to look at others.
+            parsedNodes.push({ sourceFile, body });
+        }
+    }
+    // Heuristic: fewer errors = more likely to be the right kind.
+    parsedNodes.sort((a, b) => a.sourceFile.parseDiagnostics.length - b.sourceFile.parseDiagnostics.length);
+    return parsedNodes[0].body(parsedNodes[0].sourceFile);
+}
+
+function placeNodeGroup(file: SourceFile, changeTracker: ChangeTracker, nodes: NodeArray<Node>, focusLocations?: TextSpan[][]) {
+    // 1. Grab the first node, see what kind it is, to see whether we'll need
+    //    to expand our scope.
+    // 2. Get nodes for each focusLocation by finding the specific node that
+    //    contains it, and making a comparison with the first node type.
+    // 3. Figure out how much of a range we're going to replace or insert by
+    //    using first and last node from `nodes`. Use node types and names to
+    //    match ranges.
+    if (isClassElement(nodes[0])) {
+        placeClassNodeGroup(file, changeTracker, nodes as NodeArray<ClassElement>, focusLocations);
+    }
+    else if (isTypeElement(nodes[0])) {
+        placeClassNodeGroup(file, changeTracker, nodes as NodeArray<TypeElement>, focusLocations);
+    }
+    else {
+        placeStatements(file, changeTracker, nodes as NodeArray<Statement>, focusLocations);
+    }
+}
+
+function placeClassNodeGroup(file: SourceFile, changeTracker: ChangeTracker, nodes: NodeArray<ClassElement> | NodeArray<TypeElement>, focusLocations?: TextSpan[][]) {
+    // We have one or more class-ish members.
+    //
+    // 1. If focusLocations is null/undefined, find the first class in the
+    //    SourceFile and insert/replace into that class.
+    // 2. If we have focusLocations, go over each one and:
+    //   a. Find the nearest class or interface declaration that contains the focusLocation.
+    let classOrInterface: ClassLikeDeclaration | InterfaceDeclaration | undefined;
+    if (!focusLocations || !focusLocations.length) {
+        for (const stmt of file.statements) {
+            if (isClassLike(stmt)) {
+                classOrInterface = stmt;
+                break;
+            }
+            else if (stmt.kind === SyntaxKind.InterfaceDeclaration) {
+                classOrInterface = stmt as InterfaceDeclaration;
+                break;
+            }
+        }
+    }
+    else {
+        let node: Node | undefined;
+        top: for (const locationGroup of focusLocations) {

```

Comment:
requires focusLocations to be flattened:
```suggestion
        classOrInterface = focusLocations.find(loc => findAncestor(getNodeAtPosition(original, loc.start), node => isClassLike(node) || isInterfaceDeclaration(node)));
```